### ISSUE TITLE
[S3] Backlight not responsive on KDE Neon with 5.9.6-surface

### ISSUE THREAD

**Issue by KJP12** [linux-surface/linux-surface#336]
Backlight is not responsive to any of the following:
 - Keyboard brightness controls
    - Plasma did acknowledged that the buttons were being pressed, but the screen never changed brightness)
 - Plasma's Display Brightness under Battery and Brightness from the system tray
 - Power management dimming the screen from idling
 - Suspending
    - The screen never turned off, just stayed at full brightness despite displaying black. The system did go into S0-sleep otherwise, noted by how kscreenlocker took a bit of time to show up after any external input
 - Directly setting via `echo "1" > /sys/class/backlight/acpi_video0/brightness` as root (You may put any integer between 0 and 100 in place of 1)
    - This setting is persistent and loads during boot (including Plymouth & console) unlike all the other ways listed. Booting with 5.8.15-surface shows that this setting did actually set, but didn't update the backlight under 5.9.6-surface.

Note, downgrading from 5.9.6-surface to 5.8.15-surface does allow for the backlight to function normally, indicating that the bug was introduced some time between 5.8.15 and 5.9.6, or by one of the patches.

### Environment

- Hardware model: Surface 3 (non-pro, no LTE)
- Kernel version: [Debian-packaged 5.9.6-surface](https://github.com/linux-surface/linux-surface/releases/tag/debian-5.9.6-1)
- Distribution: KDE Neon 5.20.3 (Ubuntu-based)

*No output of `dmesg` has been captured during the run with 5.9.6-surface.*

**Comment by Rev-b** [comment]
I can confirm this, with kernels 5.9.6 and 5.9.8 backlight control stopped working.

Reverting to 5.9.3 fixes it.

**Comment by qzed** [comment]
Can you check if this problem exists with upstream v5.9.6/v5.9.8? You can get Debian kernels from https://kernel.ubuntu.com/~kernel-ppa/mainline/ (you'll need to disable secureboot for that).

**Comment by KJP12** [comment]
It appears to be an upstream bug, 5.9.8-050908-generic has the same bug of the backlight not responding to any brightness controls, including suspending.

**Comment by qzed** [comment]
Can you try to figure out which kernel breaks this first using the PPA builds I've linked above? That would help us narrow it down a bit.

**Comment by KJP12** [comment]
5.9.4-050904-generic -> Backlight *not* controllable
5.9.3-050903-generic -> Backlight is controllable

**Comment by jwrdegoede** [comment]
This works fine with the mainline kernel for me on my S3. A dmesg output would really help.

I would expect:

```
dmesg | grep PWM
```

to output:

```
i915 0000:00:02.0: [drm]  *ERROR* Failed to get the PMIC PWM chip
```

Where as with a working config I would expect:

```
i915 0000:00:02.0: [drm] Using PMIC PWM for LCD backlight control
```

This error is typically caused by a kernel Kconfig / .config issue. All of the following need to be set to Y for things to work properly:

```
CONFIG_I2C_DESIGNWARE_CORE=y
CONFIG_I2C_DESIGNWARE_PLATFORM=y
CONFIG_I2C_DESIGNWARE_BAYTRAIL=y
CONFIG_I2C_DESIGNWARE_PCI=y
CONFIG_INTEL_SOC_PMIC=y
CONFIG_PWM_CRC=y
```

**Comment by jwrdegoede** [comment]
This works fine with the mainline kernel for me on my S3. A dmesg output would really help.

I would expect:

```
dmesg | grep PWM
```

to output:

```
i915 0000:00:02.0: [drm]  *ERROR* Failed to get the PMIC PWM chip
```

Where as with a working config I would expect:

```
i915 0000:00:02.0: [drm] Using PMIC PWM for LCD backlight control
```

This error is typically caused by a kernel Kconfig / .config issue. All of the following need to be set to Y for things to work properly:

```
CONFIG_I2C_DESIGNWARE_CORE=y
CONFIG_I2C_DESIGNWARE_PLATFORM=y
CONFIG_I2C_DESIGNWARE_BAYTRAIL=y
CONFIG_I2C_DESIGNWARE_PCI=y
CONFIG_INTEL_SOC_PMIC=y
CONFIG_PWM_CRC=y
```

**Comment by qzed** [comment]
@jwrdegoede Thanks! Our Debian/Ubuntu config has `CONFIG_I2C_DESIGNWARE_PCI=m`, so that may have caused the issue. I'll fix that.

**Comment by qzed** [comment]
Reopening until we have confirmation.

@KJP12 New kernels are building, could you test this once they're available?

**Comment by jwrdegoede** [comment]
I just checked and the I2C controllers on the S3 are indeed PCI enumerated rather then ACPI enumerated, so that might very well be the cause.

With the I2C driver as module there will be a race whether the I2C controller or the i915 driver probes first and if the i915 driver wins then the PWM driver will not have loaded yet and the backlight interface won't work then. 

Part of the problem here is the i915 driver not supporting -EPROBE_DEFER, so all resources from other drivers which it needs must be in place before it loads.

**Comment by qzed** [comment]
@jwrdegoede Thanks! That makes sense.

**Comment by KJP12** [comment]
It has been a bit, although, 5.12.12-surface, as I have last left it since I've done anything to the Surface 3, does have a working backlight driver; I forgot to mention it a while ago.

I will still be checking to make sure that this new release still works as expected. It may take me a bit as I am practically upgrading the entire system. I'll report back when I can install 5.16.11-surface and give a try, also giving the stock and 5.12.12 kernels a try as baseline if any between then and 5.16.11 reintroduces the same error

**Comment by KJP12** [comment]
5.16.11-surface does indeed work as expected. I'll give it a try for a bit longer as I did notice that the brightness pinned itself to 100% after it turned off & on the screen after the brightness was previously set to 0%, although, that may also have been a bug within KDE Plasma as well; it's hard to tell with an unusual level like that.

A bit unrelated tho, but the repository for the kernel & wacom library does seem to be redirecting to a GitHub repository returning 401s to anyone trying to update? Ended up manually downloading & installing the .deb files from releases instead as it doesn't seem like anything resolvable from my end.

**Comment by qzed** [comment]
> A bit unrelated tho, but the repository for the kernel & wacom library does seem to be redirecting to a GitHub repository returning 401s to anyone trying to update? Ended up manually downloading & installing the .deb files from releases instead as it doesn't seem like anything resolvable from my end.

This is an unfortunate bug in APT, triggered by some redirect-URLs from GitHub. See https://github.com/linux-surface/linux-surface/issues/625#issuecomment-962622556 for a workaround. Alternatively, this should be fixed in Ubuntu 21.04 and upwards.

> Looks like it has no different behaviour from 5.12.12-surface, so, I'd consider it fixed.

Thanks! Closing this then.
